### PR TITLE
Refactor EntitlementService and extract License from Chargebee (2/3)

### PR DESCRIPTION
**Pull request by geropl**
## Description
This again is a refactor with the aim to have 0 behavior changes. It's best reviewed in Gitpod [method by method](https://github.com/gitpod-io/gitpod/pull/11889/files#diff-da9747280402a215ed0c5b3a2a4cf352032977892ab33a1686c6b0f73019cef2) of this change.

There are two changes in call hierarchy, which makes this PR a non 100% move op:
 - before:
   - caller -> UserService.getDefaultWorkspaceTimeout -> EntitlementService.getDefaultWorkspaceTimeout
   - caller -> UserService.userGetsMoreResources -> EntitlementService.userGetsMoreResources
 - after:
   - caller -> EntitlementService.*

The behavior of UserService.* has been folded into the new one be extracting that code into `EntitlementServiceLicense`. As those `EntitlementService.*` methods have only been called by UserService, it's guaranteed to not change behavior. :lotus_position: 

## Related Issue(s)

Context: https://github.com/gitpod-io/gitpod/issues/11402

### setup
- [x] create a branch based on license: https://gpl-entitl0bde00adcc.preview.gitpod-dev.com/workspaces ([other branch](https://github.com/gitpod-io/gitpod/pull/11906))
- [x] setup payment webhook

### License
- go to https://gpl-entitl0bde00adcc.preview.gitpod-dev.com/workspaces and notice that you can do everything as allowed per license (and nothing more)

### Chargebee
- got to https://gpl-entitl342918dfed.preview.gitpod-dev.com/workspaces (this branch)
- start workspace, try to set workspace timeout => does not work
- upgrade to a paid plan, try again and notice the difference.

## Release Notes

```release-note
NONE
```

## Werft options:

- [x] /werft with-preview
- [x] /werft with-payment

### CODE REVIEW

File: components/server/ee/src/container-module.ts [hunk 1]
```diff
@@ -64,6 +62,8 @@ import { BitbucketServerApp } from "./prebuilds/bitbucket-server-app";
 import { EntitlementService } from "../../src/billing/entitlement-service";
 import { EntitlementServiceChargebee } from "./billing/entitlement-service-chargebee";
 import { BillingModes, BillingModesImpl } from "./billing/billing-mode";
+import { EntitlementServiceLicense } from "./billing/entitlement-service-license";
+import { EntitlementServiceImpl } from "./billing/entitlement-service";
```
Comment (AlexTugarev) [inline components/server/ee/src/container-module.ts:66]: now re-reading this, it seems odd that ES is in "billing", especially that EntitlementServiceImpl is delegating to different strategies.

File: components/server/ee/src/workspace/gitpod-server-impl.ts [hunk 2]
```diff
@@ -331,7 +331,7 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
 
         // if any other running instance has a custom timeout other than the user's default, we'll reset that timeout
         const client = await this.workspaceManagerClientProvider.get(runningInstance.region);
-        const defaultTimeout = await this.userService.getDefaultWorkspaceTimeout(user);
+        const defaultTimeout = await this.entitlementService.getDefaultWorkspaceTimeout(user, new Date());
```
Comment (AlexTugarev) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:334]: 👍🏻

File: components/server/src/billing/entitlement-service.ts [hunk 3]
```diff
@@ -64,10 +64,6 @@ export class CommunityEntitlementService implements EntitlementService {
         return true;
     }
 
-    async hasFixedWorkspaceResources(user: User, date: Date = new Date()): Promise<boolean> {
```
Comment (AlexTugarev) [inline components/server/src/billing/entitlement-service.ts:67]: was it a left-over?

File: components/server/ee/src/container-module.ts (hunk 1 above)
Comment (geropl) [reply components/server/ee/src/container-module.ts:66]: Jep. Thought about making it `/entitlement/` but could stomach that quickly, so left it here, for now. :shrug:

File: components/server/src/billing/entitlement-service.ts (hunk 3 above)
Comment (geropl) [reply components/server/src/billing/entitlement-service.ts:67]: :100:

File: components/server/ee/src/billing/entitlement-service-license.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        return { enoughCredits: true };
+    }
+
+    async maySetTimeout(user: User, date: Date): Promise<boolean> {
+        // when payment is disabled users can do everything
+        return true;
+    }
+
+    async getDefaultWorkspaceTimeout(user: User, date: Date): Promise<WorkspaceTimeoutDuration> {
+        const userCount = await this.userDb.getUserCount(true);
```
Comment (AlexTugarev) [inline components/server/ee/src/billing/entitlement-service-license.ts:43]: TODO (out-of-scope here) using `UserCounter` as with other occurrences of `licenseEvaluator.isEnabled`

File: components/server/ee/src/billing/entitlement-service-license.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        if (!this.licenseEvaluator.isEnabled(Feature.FeatureSetTimeout, userCount)) {
+            return WORKSPACE_TIMEOUT_DEFAULT_SHORT;
+        }
+
+        return WORKSPACE_TIMEOUT_DEFAULT_LONG;
+    }
+
+    async userGetsMoreResources(user: User): Promise<boolean> {
+        // TODO(gpl) Not sure this makes sense, but it's the way it was before
+        return false;
```
Comment (AlexTugarev) [inline components/server/ee/src/billing/entitlement-service-license.ts:55]: `return true` actually cf. https://github.com/gitpod-io/gitpod/pull/11889/files#diff-da9747280402a215ed0c5b3a2a4cf352032977892ab33a1686c6b0f73019cef2L175

Comment (geropl) [reply components/server/ee/src/billing/entitlement-service-license.ts:55]: Ya-ha, this is where it's getting creepy :grimacing: Thank you for the thorough review, exactly what I hoped for! :pray: 

This methods only direct callsite was [here](https://github.com/gitpod-io/gitpod/pull/11889/files#diff-6eac038c4219e90db377c71da65f9a2a5521b0ad013330d63d8b0ff6f10aa65aL76), which returns `false`... again, not sure this is the best decision, but want to strictly keep existing behavior in this PR.

File: components/server/ee/src/billing/entitlement-service.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import { EntitlementService } from "../../../src/billing/entitlement-service";
+import { Config } from "../../../src/config";
+import { MayStartWorkspaceResult } from "../user/eligibility-service";
+import { EntitlementServiceChargebee } from "./entitlement-service-chargebee";
+import { EntitlementServiceLicense } from "./entitlement-service-license";
+
+/**
+ * The default implementation for the Enterprise Edition (EE). It decides based on config which ruleset to choose for each call.
+ */
+export class EntitlementServiceImpl implements EntitlementService {
```
Comment (AlexTugarev) [inline components/server/ee/src/billing/entitlement-service.ts:18]: `@injectable` is missing

Comment (geropl) [reply components/server/ee/src/billing/entitlement-service.ts:18]: :scream: :100:

**Comment by AlexTugarev** [comment]
@geropl, I couldn't set workspace timeout after switching to a paid plan on 

https://alextugarev-bel-eka8ju26132.ws.gpl-entitl342918dfed.preview.gitpod-dev.com/

Screen Shot 2022-08-05 at 11 49 56

**Comment by geropl** [comment]
@AlexTugarev let's meet, what branch is that? :thinking:

File: components/server/ee/src/billing/entitlement-service-license.ts (hunk 5 above)
Comment (AlexTugarev) [reply components/server/ee/src/billing/entitlement-service-license.ts:55]: dead short

**Comment by AlexTugarev** [comment]
Thanks for clarifying. It's the Unleashed plan which is needed to enable the extended workspace timeout.

File: components/server/ee/src/billing/entitlement-service.ts (hunk 6 above)
Comment (geropl) [reply components/server/ee/src/billing/entitlement-service.ts:19]: Fixed.

**Review by AlexTugarev — APPROVED** [review]

**Pull request merged**